File: ruby/lesson 10/how_sum_recursive.rb
def how_sum(target_sum, numbers)
    how_sum_impl(target_sum, numbers, Array.new(target_sum + 1))
end

def how_sum_impl(target_sum, numbers, dp)
    return nil if target_sum.negative?
    return [] if target_sum.zero?
    return dp[target_sum] unless dp[target_sum].nil?
    
    numbers.each do |number|
        current = how_sum_impl(target_sum - number, numbers, dp)
        unless current.nil?
            current.push(number)
            dp[target_sum] = current
            return dp[target_sum]
        end
    end
    
    nil
end

File: kotlin/lesson 10/HowSumRecursive.kt
fun howSum(targetSum: Int, numbers: IntArray): List<Int>? {
    var dp = Array<MutableList<Int>?>(targetSum + 1, { null })
    return howSum(targetSum, numbers, dp)
}

fun howSum(targetSum: Int, numbers: IntArray, dp: Array<MutableList<Int>?>): MutableList<Int>? {
   if (targetSum < 0) return null
   if (targetSum == 0) return mutableListOf<Int>()
   if (dp[targetSum] != null) return dp[targetSum]

   for (number in numbers) {
       val current = howSum(targetSum - number, numbers, dp)
       if (current != null) {
           current.add(number)
           dp[targetSum] = current
           return dp[targetSum]
       }
   }

   return null
}
